src/material/form-field/form-field.ts

```diff
@@ -552,10 +554,26 @@ export class MatFormField
 
   /** Refreshes the width of the outline-notch, if present. */
   _refreshOutlineNotchWidth() {
-    if (!this._hasOutline() || !this._floatingLabel) {
-      return;
+    if (this._hasOutline() && this._floatingLabel) {
+      this._labelWidth = this._floatingLabel.getWidth();
     }
-    this._labelWidth = this._floatingLabel.getWidth();
+  }
+
+  /** Invoked when the content of the label has changed. */
+  _labelContentChanged(): void {
+    this._refreshOutlineNotchWidth();
+    // Required since the changed might have happened after check,
+    // e.g. if the `required` state of the control has changed.
+    this._changeDetectorRef.detectChanges();
```

Comment:
The control already emits to `stateChanges`, but the asterisk itself doesn't show up. I think that some of the data bindings in the form field need to be re-evaluated.

Reply:
That seems surprising to me, but I also haven't played with the form field in a while. I think the flow should be:

Control changes `required` -> emits state changes -> form field is marked for check -> on re-render, asterisk should render -> content change causes outline refresh (via cdk observe content).

I don't understand why we need an explicit `detectChanges` here (and no `markForCheck`), and also why this is needed at all. Just trying to make sure we understand it before forcing an explicit change detection. Although it's likely fair to say that the content of a label doesn't change often.

Reply:
What would be the problem with waiting for the next change detection? You mean, that might not happen "soon" because nothing asynchronous is triggering Zone/a new application tick? If so, that seems reasonable to me

Reply:
Yeah so the problem is that it may take until the next user interaction for it get picked up. This could be problematic if `required` is based on a data binding.

Reply:
Gotcha. Good that we have this captured here. Ideally the comment would reflect some of these comments, but the PR is also sufficient I think.